packages/react-form-state/README.md

```diff
@@ -0,0 +1,979 @@
+- [redux-form](https://redux-form.com/7.4.2/) offers a redux integrated solution that can be quite magical and gives you the power of the redux dev tools for debugging broken states. Not all react apps need or want redux, and the library is fairly heavy, so we found it unsuitable for our uses.
+
+As such the main difference in our solution is the explicit, declarative api. Form fields are given explicit handlers generated individually for each field and made available through a [render prop](https://reactjs.org/docs/render-props.html). To reduce boilerplate we've kept the generated `field` objects as easy to use as possible. Usually you can just splat them onto your inputs. We'll go into this in more detail in the [usage](#Usage) section.
+
+Creating handlers for each field instead of having one that you share or inject using a higher order component means we are fully type safe, with editor autocomplete able to show a developer exactly what fields are available, and compilation breaking when someone tries to use a field that doesn't exist.
+
+## Installation
+
+```bash
+$ yarn add @shopify/react-form-state
+```
+
+## Usage
+
+### Basic
+
+The default component exported by this library is `<FormState />`.
+
+```typescript
+import FormState from '@shopify/react-form-state';
+```
+
+```typescript
+interface Props<Fields> {
+  initialValues: Fields;
+  validators?: Partial<ValidatorDictionary<Fields>>;
+  onSubmit?: SubmitHandler<Fields>;
+  children(form: FormDetails<Fields>): React.ReactNode;
+}
+```
+
+It's only mandatory props are `initialValues` and `children`. The `initialValues` prop is used to infer all the types for the rest of the component, and to generate handlers and field state objects. The `children` prop expects a function of the current state of the form, which is represented by a `FormDetails` object.
+
+```typescript
+<FormState initialValues={myInitialValues}>
+  {({fields, dirty, valid, submitting, errors, reset, submit}) => {
+    return /* some cool ui */;
+  }}
+</FormState>
+```
+
+When using `<FormState>`, you express your form's component tree as a function of the generated field state objects provided by `<FormState />` into it's `children` function as `formDetails`.
+
+```typescript
+export interface FormDetails<Fields> {
+  fields: FieldDescriptors<Fields>;
+  dirty: boolean;
+  valid: boolean;
+  submitting: boolean;
+  errors: ClientError[];
+  reset(): void;
+  submit(): void;
+}
+```
+
+The `formDetails` object passed into `children` contains a `fields` dictionary, which includes all the state and handlers you need to render your form.
+
+```typescript
+export interface FieldState<Value> {
+  name: string;
+  initialValue: Value;
+  value: Value;
+  dirty: boolean;
+  changed: boolean;
```

Comment:
👍 Happy to see you added the `changed` prop :)

Reply:
This is actually out of date I had removed it because it's not yet useful (we're not doing full form validation on each blur / update yet)